File: generated_code/python/solution_2079_2.py
from typing import List, Dict
from collections import defaultdict

class TrieNode:
    def __init__(self):
        # Maps folder name to its corresponding TrieNode child
        self.children: Dict[str, TrieNode] = {}
        # Stores the serialized sub-tree representation of this node
        self.serial: str = ""
        # Flag to mark if this folder/node should be deleted
        self.deleted: bool = False

class Solution:
    def deleteDuplicateFolder(self, paths: List[List[str]]) -> List[List[str]]:
        root = TrieNode()
        
        # 1. Build the Trie from the given paths.
        # Sorting ensures that children are processed in lexicographical order later.
        for path in sorted(paths):
            curr = root
            for folder in path:
                if folder not in curr.children:
                    curr.children[folder] = TrieNode()
                curr = curr.children[folder]
                
        # Map to store the frequency of each unique sub-tree serialization
        serial_count = defaultdict(int)
        # Map to store all nodes that share a particular sub-tree serialization
        serial_to_nodes = defaultdict(list)
        
        # 2. Post-order DFS to serialize the sub-tree structure of each node.
        def serialize(node: TrieNode) -> str:
            if not node.children:
                return ""
            
            # Since paths were sorted, iterating over keys directly preserves order.
            # Otherwise, sorted(node.children.keys()) would be required.
            sub_serials = []
            for name, child in node.children.items():
                child_serial = serialize(child)
                sub_serials.append(f"{name}({child_serial})")
                
            # Combine the serialized sub-trees of all children
            node.serial = "".join(sub_serials)
            
            # Track serialization frequencies for non-empty sub-trees
            serial_count[node.serial] += 1
            serial_to_nodes[node.serial].append(node)
            
            return node.serial
            
        serialize(root)
        
        # 3. Mark duplicate folders for deletion.
        # Folders are identical if they have the same non-empty sub-tree structure.
        for serial, count in serial_count.items():
            if count > 1:
                for node in serial_to_nodes[serial]:
                    node.deleted = True
                    
        ans = []
        
        # 4. Deep DFS to reconstruct valid paths that haven't been deleted.
        def construct_paths(node: TrieNode, current_path: List[str]):
            # If a folder is marked deleted, its entire sub-tree is pruned/ignored.
            if node.deleted:
                return
                
            if current_path:
                ans.append(list(current_path))
                
            for name, child in node.children.items():
                current_path.append(name)
                construct_paths(child, current_path)
                current_path.pop()
                
        construct_paths(root, [])
        return ans

File: generated_code/cpp/solution_2079_2.cpp
#include <vector>
#include <string>
#include <unordered_map>
#include <map>
#include <algorithm>

using namespace std;

// Trie node structure to represent the file system hierarchy
struct TrieNode {
    string name;
    // Use std::map to automatically keep children sorted by name.
    // This simplifies generating consistent serialized sub-tree strings.
    map<string, TrieNode*> children;
    bool del = false; // Flag to mark if this folder should be deleted

    TrieNode(string n = "") : name(n) {}
};

class Solution {
private:
    // Global frequency map for sub-tree serializations
    unordered_map<string, int> seen;

    // Helper to insert a path into the Trie
    void insert(TrieNode* root, const vector<string>& path) {
        TrieNode* curr = root;
        for (const string& s : path) {
            if (curr->children.find(s) == curr->children.end()) {
                curr->children[s] = new TrieNode(s);
            }
            curr = curr->children[s];
        }
    }

    // Step 1: Serialize the sub-tree structures and count frequencies.
    // An empty or leaf folder returns an empty string because it has no subfolders.
    string serialize(TrieNode* node) {
        if (node->children.empty()) return "";

        string sub_tree = "";
        for (auto& pair : node->children) {
            sub_tree += "(" + pair.first + serialize(pair.second) + ")";
        }

        // Only track frequencies of folders that actually contain subfolders
        seen[sub_tree]++;
        return sub_tree;
    }

    // Step 2: Mark duplicate sub-trees for deletion
    void mark_duplicates(TrieNode* node) {
        if (node->children.empty()) return;

        string sub_tree = "";
        for (auto& pair : node->children) {
            sub_tree += "(" + pair.first + serialize(pair.second) + ")";
        }

        // If this exact sub-tree structure has been seen more than once,
        // mark this node for deletion.
        if (seen[sub_tree] > 1) {
            node->del = true;
        }

        for (auto& pair : node->children) {
            mark_duplicates(pair.second);
        }
    }

    // Step 3: Walk the Trie to collect all paths that survived deletion
    void collect_paths(TrieNode* node, vector<string>& current_path, vector<vector<string>>& result) {
        if (node->del) return; // If marked for deletion, stop traversing this branch

        if (!current_path.empty()) {
            result.push_back(current_path);
        }

        for (auto& pair : node->children) {
            current_path.push_back(pair.first);
            collect_paths(pair.second, current_path, result);
            current_path.pop_back(); // Backtrack
        }
    }

    // Helper to free allocated memory
    void clear(TrieNode* node) {
        for (auto& pair : node->children) {
            clear(pair.second);
        }
        delete node;
    }

public:
    vector<vector<string>> deleteDuplicateFolder(vector<vector<string>>& paths) {
        TrieNode* root = new TrieNode();
        seen.clear();

        // Build the Trie structure
        for (const auto& path : paths) {
            insert(root, path);
        }

        // Post-order serialization to populate the frequency map
        serialize(root);

        // Pre-order / structural traversal to mark duplicates
        mark_duplicates(root);

        // Collect all valid paths
        vector<vector<string>> result;
        vector<string> current_path;
        collect_paths(root, current_path, result);

        // Clean up memory
        clear(root);

        return result;
    }
};